zap/src/output/luau/client.rs

```diff
@@ -817,13 +924,121 @@ impl<'src> ClientOutput<'src> {
 			match ev.call {
 				EvCall::SingleSync | EvCall::SingleAsync => self.push_return_setcallback(ev),
 				EvCall::ManySync | EvCall::ManyAsync => self.push_return_on(ev),
+				EvCall::Polling => self.push_iter(ev),
 			}
 
 			self.dedent();
 			self.push_line("},");
 		}
 	}
 
+	fn push_polling(&mut self) {
+		let filtered_evdecls = self
+			.config
+			.evdecls
+			.iter()
+			.filter(|evdecl| evdecl.from == EvSource::Server && evdecl.call == EvCall::Polling)
+			.collect::<Vec<&EvDecl>>();
+
+		if !filtered_evdecls.is_empty() {
+			self.push("\n");
+		}
+
+		for evdecl in filtered_evdecls {
+			let id = evdecl.id;
+			let mut return_names: Vec<String> = vec![];
+			if evdecl.data.is_empty() {
+				return_names.push(String::from("value"));
+			}
+			for (index, parameter) in evdecl.data.iter().enumerate() {
+				return_names.push(format!(
+					"value_{}{}",
+					index + 1,
+					if let Some(name) = parameter.name {
+						format!("_{}", name)
+					} else {
+						String::from("")
+					}
+				));
+			}
+			let arguments_size = return_names.len().max(1);
+
+			self.push_line(&format!("polling_queues[{id}] = {{"));
+			self.indent();
+
+			self.push_line(&format!(
+				"arguments = table.create({}),",
+				max(arguments_size * super::INITIAL_POLLING_EVENT_CAPACITY, 1)
+			));
+			self.push_line(&format!(
+				"queue_size = {},",
+				max(arguments_size * super::INITIAL_POLLING_EVENT_CAPACITY, 1)
+			));
+			self.push_line("read_cursor = 1,");
+			self.push_line("write_cursor = 1,");
+			self.push_line("iterator = function()");
+			self.indent();
+
+			self.push_line(&format!("local queue = polling_queues[{id}]"));
+			self.push_line("local index = 0");
+			self.push_line("return function()");
+			self.indent();
+
+			self.push_line("index += 1");
+
+			self.push_line("if queue.read_cursor == queue.write_cursor then");
+			self.indent();
+
+			self.push_line("return nil");
+
+			// end `if queue.read_cursor == queue.write_cursor then`
+			self.dedent();
+			self.push_line("end");
+
+			// `queue` fields could be changed if the user yields while polling and more events are written that force the queue to grow.
+			// That's why these are defined inside the generator.
+			self.push_line("local arguments = queue.arguments");
+			self.push_line("local read_cursor = queue.read_cursor");
+			self.push_line("local queue_size = queue.queue_size");
+			self.push_indent();
+			self.push("local ");
+			self.push(&return_names.join(", "));
+			self.push(" = ");
+			for argument_index in 0..return_names.len() {
+				if argument_index > 0 {
+					self.push(", ");
+				}
+				self.push(&format!(
+					"arguments[{}]",
+					if argument_index > 0 {
+						format!("((read_cursor + {} - 1) % queue_size) + 1", argument_index)
+					} else {
+						String::from("read_cursor")
+					}
```

Comment:
Two concerns:

1. isn't reading from `read_cursor` reading the value from before where it should be (with the `+1`)
2. Let's add a fast call (`else if`) if `i == 1` so we don't get `queue.read_cursor = ((queue.read_cursor + 1 - 1) % queue.queue_size) + 1`